### PR TITLE
First draft of tuples lesson

### PR DESCRIPTION
**Pull request by mdippery**
One note: I left out the bit about modifying the `downloadImage(atURL:)` function to return a tuple with named elements. In my testing, I couldn't get this to work:

``` swift
func downloadImage(atURL url: String) -> (Bool, String) {
    return (success: false, errorMessage: "Image no longer exists")
}

let result = downloadImage(atURL: "http://example.com/image.png")
print("Success? \(result.success)")
print("Message: \(result.errorMessage)")
```

Maybe I did something wrong, but when I tried that out in a playground, Xcode wouldn't let me do `result.success` and `result.errorMessage`—it would only let me do `result.0` and `result.1`. I _think_ maybe named fields don't survive outside of the scope they're in? Or Xcode can't infer named fields from a function declaration, and named fields don't actually carry through as part of the data structure? I'd have to delve into the inner workings of tuples to confirm, but I imagine that's the case. Regardless, I figured it could be left out. I talk about named fields elsewhere and it didn't seem important to force it with this example.

I like the `downloadImage(atURL:)` example, though. I think this can be neatly revisited in the lesson on structs or enums. The Haskell/Scala programmer in me pretty quickly saw that this would be better represented using something like Haskell's and Scala's `Either` type, which could be mimicked in Swift like this:

``` swift
enum Either {
    case Left(String)
    case Right
}

func downloadImage(atURL url: String) -> Either {
    return .Left("Image no longer exists")
}

let result = downloadImage(atURL: "http://example.com/image.png")
if case .Left(let errorMsg) = result {
    print("Could not download image: \(errorMsg)")
} else {
    print("Image downloaded!")
}
```

So I think that could be a way to revisit this lesson later on.

cc: @JimCampagno @AnnJohn

### CODE REVIEW

**Comment by JimCampagno** [comment]
Reviewing this lab now. But to answer your initial question first. It appears naming the parameters in the method signature allows for you to access the members of the tuple by name when calling on that function. Similar to the ability to name the parameters of a function.. where that function is the argument to another function. As far as what's going on under the hood... not sure.

``` swift
func downloadImage(atURL url: String) -> (success: Bool, errorMessage:  String) {
    // TODO: Download image
    // if no error, return true
    // otherwise, get error string and return false
    let error = "Image no longer exists"
    return (false, error)
}

let result = downloadImage(atURL: "http://example.com/image.png")
print("Success? \(result.success)")
print("Message: \(result.errorMessage)")
```

---

example

---

anotherExample

---

enumFun

@mdippery

File: README.md [hunk 1]
```diff
@@ -1,171 +1,192 @@
 # Tuples
 
-<img src="http://i.imgur.com/gMMSTqF.jpg?1" alt="Drawing" style="width: 200px;"/>  
-
+![Drawing](http://i.imgur.com/gMMSTqF.jpg?1)
```
Comment (JimCampagno) [inline README.md:3]: Thanks for fixing that for me. I don't know why I originally ran with the above for a while. I had fixed that in later lessons.

File: README.md [hunk 2]
```diff
@@ -1,171 +1,192 @@
...
+<!-- I'm not sure how necessary this section is, given that a more practical example of tuples is given below. I thought it was a nice intro, but I'm okay with deleting it entirely, too. -->
+
+Consider a relatively simple data structure: A point in 2D space. You probably remember from geometry that such a point consists of two coordinates _(x,y)_ which together tell you where a point lies in a graph. Each individual value—_x_ and _y_—are important, but when you think of a point, you really think of these values together as _one unit_. An _x_ alone alone won't describe the location of a point. Nor will a _y_ value alone. You need both—together—to know where a point lies.
+
+What data structure would you use?
+
+You could use an array like you learned about in a previous unit, but an array doesn't really capture the _spirit_ of a 2D point. An array is more like a _list of things_. A 2D point is something special...it consists of multiple values, but those values are tied inextricably tied together and treated as a single unit.
+
+You could instead use another Swift data structure to represent a 2D point: a _tuple_.
+
```
Comment (JimCampagno) [inline README.md:25]: clap

File: README.md [hunk 3]
```diff
@@ -1,171 +1,192 @@
...
+## Much Ado About Tuples
+
+Geometrical points are pretty abstract, so let's take a look at a more practical example. Say you were tasked with writing a function called `downloadImage(AtURL:)` that will download an image at a given URL, and return `true` if the image was downloaded. Your function might look something like this:
```
Comment (JimCampagno) [inline README.md]: JIM FIX: 

`downloadImage(atURL:)`

File: README.md [hunk 4]
```diff
@@ -1,171 +1,192 @@
...
+Great! Err...not so great, actually. The function is supposed to return a `Bool`, but now it is returning a `String`, and that's not possible in Swift. That code won't even compile! And even if it could, you'd still like to also return `false` if the image download fails. Really, you need to be able to return two _values_ from a function: One of type `Bool`, indicating success or failure of the function, and one of type `String`, with a message indicating why the function failed (and some other positive affirmation if the function succeeds).
+
+But how can you do that?
+
+When you think about it, what you're trying to do is a little bit like representing a 2D point: You want to have _two_ values that are treated as _one_ unit (since a function can only return _one_ thing).
+
+You want to return a tuple.
+
+In Swift, a tuple is a funny word (seriously, it doesn't exactly [roll off the tongue](https://www.youtube.com/watch?v=7fTVCQRQYXk)) that describes a structure in which multiple values are grouped together as one. Like a 2D point! Or like returning a boolean and an error message as one unit.
+
```
Comment (JimCampagno) [inline README.md:71]: Perfect.

File: README.md [hunk 5]
```diff
@@ -1,171 +1,192 @@
...
-* You can think of a tuple as being a container of other types except it represents one value that could be stored inside of a variable. Well what does that mean?
+Take particular notice of the return type. It changed from `Bool` to `(Bool, String)`. That kind of looks almost like a function argument list, doesn't it? Well, looks can be deceiving. A tuple doesn't really have anything to do with a function argument list; it just shares a similar syntax. It's a type unto itself, just like `Bool` and `String` and `Int` are types.
+
+Notice, too, how the `return` statement changed. It says `return (false, error)` now. That's how you create a tuple: By putting all of its values in parentheses, separated by commas. Pretty easy, right? In fact, it doesn't look that much different than the way you wrote out 2D points in geometry class—because it's not really that different at all!
```
Comment (JimCampagno) [inline README.md:103]: Your approach in this entire section was very good.

File: README.md [hunk 6]
```diff
@@ -1,171 +1,192 @@
...
+print("\(person2.name)'s age is \(person2.age)")
+// prints "Jim's age is 30"
+person2.age = 32
+print("\(person2.name)'s age is \(person2.age)")
+// prints "Jim's age is 32"
 ```
 
-* In this code above, we're checking to see if the image we're attempting to download was successful by checking the `Bool` value stored in the `imageDownload` constant. If `imageDownload.success` is `true` **THEN** display the image to the user **ELSE** (if the value of `imageDownload.success` is `false`) then display the error message to the user. We're also printing the message to the console by passing in `imageDownload.errorMessage` to the `print()` function as `imageDownload.errorMessage` is of type `String`.
+Often times, tuples are constant. They're basically just containers for data, and that data doesn't generally change once a tuple has been created. But now you know that you _can_ change the elements if you want to.
+
```
Comment (JimCampagno) [inline README.md:185]: This section was perfect.

**Comment by JimCampagno** [comment]
This one is great. This entire thing is starting to take form and I think it will be very special. In part, I have you to thank for a lot of that.. so thank you.

@mdippery

**Comment by mdippery** [comment]
You're welcome! The outlines have been very helpful, too.

I read your note about returning named tuples from functions, which makes sense—I hadn't tried making it part of the function signature itself.

**Pull request merged**